library/waku_thread/inter_thread_communication/requests/peer_manager_request.nim

```diff
@@ -7,27 +7,29 @@ import
   stew/results,
   stew/shims/net
 import
-  ../../../waku/node/waku_node,
-  ./request
+  ../../../../waku/node/waku_node
 
 type
   PeerManagementMsgType* = enum
     CONNECT_TO
 
 type
-  PeerManagementRequest* = ref object of InterThreadRequest
+  PeerManagementRequest* = object
     operation: PeerManagementMsgType
     peerMultiAddr: string
     dialTimeout: Duration
 
 proc new*(T: type PeerManagementRequest,
           op: PeerManagementMsgType,
           peerMultiAddr: string,
-          dialTimeout: Duration): T =
+          dialTimeout: Duration): ptr PeerManagementRequest =
 
-  return PeerManagementRequest(operation: op,
-                               peerMultiAddr: peerMultiAddr,
-                               dialTimeout: dialTimeout)
+  var ret = cast[ptr PeerManagementRequest](
+                      allocShared0(sizeof(PeerManagementRequest)))
+  ret[].operation = op
+  ret[].peerMultiAddr = peerMultiAddr
```

Comment:
`string` is a garbage-collected type - here, a copy must be taken with the shared allocator (and later it must be deallocated) - no garbage collected types allowed in objects constructed with `create`: https://status-im.github.io/nim-style-guide/interop.html#garbage-collected-types

Reply:
(type needs changing to `cstring`)

Reply:
Okay, this is the trickiest part. 

I was assuming that a thread-safe communication was ensured by just sending `ptr` types, given that the `ptr` type is not tracked by the GC by default. I also wanted to avoid the overhead of serializing/parsing JSON objects.

My assumption was that the next was secure:
1. `Thread A` creates a `ptr` of the request in the shared space.
2. `Thread A` sends the address of the request object to `Thread B`.
3. `Thread B` handles the request and deallocates the memory from the shared space.

The [next](https://github.com/waku-org/nwaku/blob/78ebb3af77bdeca1d62761f79ecb2c3b26c08728/library/waku_thread/waku_thread.nim#L23C58-L23C58) is used to communicate both threads:
```
    reqChannel: ChannelSPSCSingle[ptr InterThreadRequest]
    respChannel: ChannelSPSCSingle[ptr InterThreadResponse]
```

The `RelayRequest` type is the most complex example:
```
type
  RelayRequest* = object
    operation: RelayMsgType
    pubsubTopic: PubsubTopic
    relayEventCallback: WakuRelayHandler
    message: WakuMessage
```
... which is created by `Thread A` in https://github.com/waku-org/nwaku/blob/78ebb3af77bdeca1d62761f79ecb2c3b26c08728/library/waku_thread/inter_thread_communication/requests/protocols/relay_request.nim#L34

... and deallocated by `Thread B` in: https://github.com/waku-org/nwaku/blob/78ebb3af77bdeca1d62761f79ecb2c3b26c08728/library/waku_thread/inter_thread_communication/requests/protocols/relay_request.nim#L44

Isn't that a thread-safe approach to sending `ptr` types over?

Reply:
`peerMultiAddr` is a `string` which is a garbage-collected type (together with seq, ref and closures). Both the "main object" and all its fields need to be thread safe, ie non-req.

This means that we manually allocate a copy of everything on `createShared` and release it - I suggest implementing a `destroyShared` function for every `createShared` which deallocates all fields and finally the main object - in the future, these can be turned into proper destructors but they are good for now.